[MINOR] Add TypeReference and TypeName fields

nickwilliams-eventbrite opened the pull request:
Add fields that can validate a type (optionally a subclass of another type) or a string that can be resolved to a type.

vartec commented:
Code looks good to me. 

My concern is however that until now Conformity validations would work independently of the context they're executed in. Validating if type is importable on the other hand strongly depends on the environment you're checking in. This will give different results depending on if module is installed or not.

nickwilliams-eventbrite commented:
> My concern is however that until now Conformity validations would work independently of the context they're executed in. Validating if type is importable on the other hand strongly depends on the environment you're checking in. This will give different results depending on if module is installed or not.

This is true. However, it will be highly useful for things like configuration. PySOA already does a hacked-together form of this for configuring transports, metrics, etc., in its Conformity-based configuration schemas, and I'm looking at using Conformity for configuration validation in an internal tool of ours, which would also be able to use this.

nickwilliams-eventbrite commented:
> This looks good to me. The only thing I might change is the naming for `TypeName`. Maybe `TypePath`?

Renamed `TypeName` to `TypePath` and `type_name` to `type_path`, and added some more assertions in the tests.
